Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp

```diff
@@ -91,24 +95,55 @@ std::optional<CryptoKeyPair> CryptoKeyOKP::platformGeneratePair(CryptoAlgorithmI
         return std::nullopt;
     }
 
-    // Retrieve the `q` data, which should be in the uncompressed point format for the X25519 keys.
-    // Validate the data size and the first byte (which should be 0x40).
-    size_t pubKeySize = namedCurve == NamedCurve::Ed25519 ? 32 : 33;
-    auto publicKeyData = mpiData(qMpi);
-    if (!publicKeyData || publicKeyData->size() != pubKeySize
-        || (namedCurve == NamedCurve::X25519
-            && !CryptoConstants::matches(publicKeyData->data(), 1, CryptoConstants::s_x25519UncompressedFormatLeadingByte))) {
+    auto q = mpiData(qMpi);
+    auto d = mpiData(dMpi);
+    if (UNLIKELY((!q || !d)))
+        return std::nullopt;
+    return std::make_pair(WTFMove(*q), WTFMove(*d));
+}
+
+static std::optional<std::pair<Vector<uint8_t>, Vector<uint8_t>>> gcryptGenerateX25519Keys()
+{
+    // private key is just 32 random bytes
+    PAL::GCrypt::Handle<gcry_mpi_t> mpi(gcry_mpi_new(256));
+    gcry_mpi_randomize(mpi, 256, GCRY_STRONG_RANDOM);
+    auto q = mpiData(mpi);
+    if (UNLIKELY(!q))
+        return std::nullopt;
+
+    // public key being X25519(a, 9), as defined in [RFC7748], section 6.1.
+    auto d = X25519(*q, WebCore::CryptoKeyOKPImpl::nine);
+    if (UNLIKELY(!d))
+        return std::nullopt;
+
+    return std::make_pair(WTFMove(*q), WTFMove(*d));
+}
+
+std::optional<CryptoKeyPair> CryptoKeyOKP::platformGeneratePair(CryptoAlgorithmIdentifier identifier, NamedCurve namedCurve, bool extractable, CryptoKeyUsageBitmap usages)
+{
+    std::optional<std::pair<Vector<uint8_t>, Vector<uint8_t>>> keyPair;
+    switch (namedCurve) {
+    case NamedCurve::Ed25519:
+        keyPair = gcryptGenerateEd25519Keys();
+        break;
+    case NamedCurve::X25519:
+        keyPair = gcryptGenerateX25519Keys();
+        break;
+    default:
         return std::nullopt;
     }
 
-    auto privateKeyData = mpiData(dMpi);
-    if (!privateKeyData || privateKeyData->size() != 32)
+    ASSERT(!!keyPair);
```

Comment:
Sorry, I was completely wrong here. Of course keyPair can still be nullopt, if either function failed and returned nullopt. Please fix before landing.

Reply:
oh, right ! sorry for not noticing it before. 
Will fix it before landing.